_posts/2023-01-07-datafusion-16.0.0.md

```diff
@@ -0,0 +1,289 @@
+and [Polars](https://www.pola.rs/)https://www.pola.rs/), DataFusion is
+closing the gap quickly. Performance highlights from the last three
+months:
+
+* XX% Faster Sorting and Merging using the new [Row Format](https://arrow.apache.org/blog/2022/11/07/multi-column-sorts-in-arrow-rust-part-1/)
+* [Advanced predicate pushdown](https://arrow.apache.org/blog/2022/12/26/querying-parquet-with-millisecond-latency/), directly on parquet, optionally directly from object storage, enabling sub millisecond filtering, directly from object storage <!-- Andrew nots: we should really get this turned on by default -->
+* Improved `IN` expressions significantly faster  Simplify InListExpr ~20-70% Faster ([#4057])
+* Sort and partition aware optimizations such as  #3969 and  #4691,  skipping potentially expensive operations
+* Basic filter selectivity analysis (#3868)
+
+
+In the coming few months, we plan work on:
+* Improved grouping performance (TODO link)
+* bloom filtering
+* investigate RLE (Run End Encoding support) (todo Arrow link)
+* Enable predicate pushdown by default for all cases
+* OTHERS?
+
+## Runtime Resource Limits
+
+Initially, DataFusion could potentially use unbounded amounts of memory for certain queries that included Sorts, Grouping or Joins.
+
+In version 16.0.0, it is possible to limit DataFusion's memory usage for Sorting and Grouping. We are looking for help adding similar limiting for Joins as well as expanding our algorithms to spill to secondary storage, if available. See #3941 fore more detail.
+
+
+## SQL Window Function
+[SQL window functions](https://en.wikipedia.org/wiki/Window_function_(SQL))  are useful for a variety of analysis and DataFusion's implementation is close to complete now.
+
+- Custom window frames such as `... OVER (ORDER BY ... RANGE BETWEEN 0.2 PRECEDING AND 0.2 FOLLOWING)`
+- Unbounded window frames such as `... OVER (ORDER BY ... RANGE UNBOUNDED ROWS PRECEDING)`
+- Support for the `NTILE` window function (#4676)
+- Support for `GROUPS` mode (#4155)
+
+
+# Improved Join Support
+
+Joins are often the most complicated operations to handle well in
+analytics systems and DataFusion 16.0.0 offers significant improvements
+such as
+
+- Cost based optimizer (CBO) that can automatically reorder join evaluations, algorithm (Merge / Hash), and pick build side based on available statistics join type (`INNER`, `LEFT`, etc) (#4219)
+- Fast non `column=column` equijoins such as `JOIN ON a.x + 5 = b.y`
+- Better optimized non-equijoins (nested loops rather than cross join and filter) (#4562) <!-- TODO is this a good thing to mention as any time this is usd the query is going to go slow or the data size is small -->
+
+# Streaming Execution
+
+TODO: write up some supporting text introducing this concept and explaining that DataFusion can now be used for stream processing, much like [Spark Streaming](https://www.databricks.com/glossary/what-is-spark-streaming)
```

Comment:
@ozankabak perhaps you can help fill in this section -

Reply:
Sure, will work on it tomorrow.

Reply:
I put the following together, feel free to enrich/fix:

--

One of the use cases for Datafusion is to serve as a foundational layer for high level streaming-first data platforms. An important prerequisite for this goal is support incremental execution for queries that allow incremental execution in theory.

With this release, significant progress has been made on this front. Datafusion now supports:
- Data ingestion from infinite files such as FIFOs (#4694),
- Detection of pipeline-breaking queries in streaming use cases (#4694),
- Rearranging join sides to support incremental execution when only one input side is a data stream (#4694),
- Intelligent elision of pipeline-breaking sort operations whenever possible (#4691),
- Incremental execution for more types of queries; e.g. queries involving finite window frames (#4777).

Stay tuned; even more improvements are on the way!

Reply:
Thanks -- this is awesome -- I incorporated in 4641b93356a